Add double router delete on router migration

AbelNavarro opened the pull request:
On cloud6 environments with DVR enabled, the router is not properly
unbound from the L3 agent and requires a double delete to have it in a
clean state. This seems to be fixed from Mitaka version onward so this
workaround is needed for Liberty only. It should not break compatibility
with subsequent versions.

AbelNavarro committed:
Add double router delete on router migration

On cloud6 environments with DVR enabled, the router is not properly
unbound from the L3 agent and requires a double delete to have it in a
clean state. This seems to be fixed from Mitaka version onward so this
workaround is needed for Liberty only. It should not break compatibility
with subsequent versions.

This is a workaround for bsc#1016943

AbelNavarro commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) S55P5 (S54P1) bsc#1016943 - neutron-ha-tool: migrating dvr router fails on first try](https://trello.com/c/D8MqR7mr/264-3-s55p5-s54p1-bsc-1016943-neutron-ha-tool-migrating-dvr-router-fails-on-first-try)

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -654,10 +654,14 @@ def migrate_router(qclient, router_id, agent, target,
-        raise RuntimeError("Failed to remove router_id=%s from agent_id=%s" %
-                           (router_id, agent['id']))
+        qclient.remove_router_from_l3_agent(agent['id'], router_id)
+        LOG.debug("Removed (again) router from agent=%s" % agent['id'])
```
Not sure that I like the workaround : ). Actually the resource agent is retrying that, and when I did some manual test this worked. So that means that if I am not wrong, the first delete raised and exception somewhere.

But if it is not raising anything, I have two comments:

* Add a reference of the bsc# bug report in the comment in the code and in the commit. Be clear in the code that this is a workaround that needs to be reverted once the real bug is fixed (or SOC6 dies)

* I wonder if "Removed (again)" is a bit misleading here. Maybe a WARNING, instead of a DEBUG and with a different text, like "The router was not correctly deleted, retrying" or something

rhafer replied:
> Actually the resource agent is retrying that, and when I did some manual test this worked. So that means that if I am not wrong, the first delete raised and exception somewhere.

During the upgrade we're not calling neutron-ha-tool from the resource agent. And the behaviour of ```--retry``` actually interferes with other requirements of the upgrade as it would also retry in cases where we need to raise an error during the ugprade (see https://bugzilla.suse.com/show_bug.cgi?id=1020914 that's why we also introduced: https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/20 ).

AbelNavarro replied:
@aplanas I don't like the workaround neither but the fix seems far down the road, because heavy backport is needed.

The retry mechanism you mention does not solve the problem here. The first (and incomplete) delete does not produce an error thus the retry mechanism is not activated.

About your two comments:
- I'll add a reference to the bsc#, good point.
- I can change the output on the retry. I like the new text you proposed. I don't think though it's a warning since it's expected behavior for cloud6.

rhafer reviewed (changes requested):
I would really prefer if the 2nd delete would only ever be invoked for distributed routers.

I guess the best way to implement that is to pass the whole router object (instead of just the router_id) into migrate_router/migrate_router_safely and checking for ```router['distributed']```.

This will of course require a bit of refactoring to neutron-ha-tool but it shouldn't be too hard and it would avoid the overhead of an additional query to the neutron API.

AbelNavarro commented:
@rhafer it won't harm to do it only for distributed routers. Will work on that.

AbelNavarro commented:
Fixed problem with variable. Tested successfully on cloud6 (manual).

rhafer commented on files/default/neutron-ha-tool.py:
```diff
@@ -623,7 +623,8 @@ def migrate_router_safely(qclient, noop, router_id, agent, target,
 
     try:
-        migrate_router(qclient, router_id, agent, target,
+        router = get_router(qclient, router_id)
```
I would really prefer if we could avoid  this additional API queries. ```list_routers_on_l3_agent``` already gets all the required attributes as a result. So if we could change it to return a list of routers instead of just router_ids it should be possible. (IIRC all the callers of ```migrate_router_safely``` process results ```list_routers_on_l3_agent```)

AbelNavarro replied:
last commit should address your request.

AbelNavarro committed:
Refactor router retrieval to lower API calls

We were calling the API many times to get the routers and then again for
the router IDs. Now the router object will be stored in memory and the
IDs retrieved from there.

rhafer commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -41,6 +41,11 @@ def list_agents(self):
             'agents': self.fake_neutron.agents.values()
         }
 
+    def show_router(self, router_id):
```
I believe this is no longer needed.

AbelNavarro replied:
good catch, removing.

rhafer commented on files/default/neutron-ha-tool.py:
```diff
@@ -615,63 +614,72 @@ def migrate_l3_routers_from_agent(qclient, agent, targets, agent_picker,
-                           (router_id, agent['id']))
+    # ensure it is removed
+    if router in list_routers_on_l3_agent(qclient, agent['id']):
+        if router['distributed']:
```
To avoid confusion on neutron releases that don't suffer from bsc#1016943 (i.e. mitaka and newer) could we please check if the router is still present before trying to delete it a second time?

AbelNavarro replied:
We actually check that. The `list_routers_on_l3_agent` contacts the API to get the latest information, so I guess that should be enough.

rhafer replied:
Ah, sorry. I misread the code :( . Forget what I said here.

rhafer reviewed (changes requested):
Apart from the two comments, I think it would be better to reorganize/squash the commits a bit. E.g. one commit that implements the router_id -> router change for ```list_routers_on_l3_agent``` and a 2nd one for the dvr double deletion.

AbelNavarro commented:
@rhafer about rebasing, I totally agree with that. That's for me the last step since it's easier to code-review when changes are incremental. So let me delay that part.

rhafer commented:
> about rebasing, I totally agree with that. That's for me the last step since it's easier to code-review when changes are incremental. So let me delay that part.

Ok. From my POV the code is good now. Apart from the rebase.

AbelNavarro commented:
rebased, squashed to two commits.

AbelNavarro commented:
waiting for milestone to be released before merging.

aplanas reviewed (approved):
Mostly good ok, but I wonder if the order of the commits are OK. For example, in one commit `get_router` is added and in the next one is removed after the refactor.

I am OK with that if @rhafer agrees, but I would prefer a more natural approach: first refactor, then implement new feature

matelakat commented on files/default/neutron-ha-tool.py:
```diff
@@ -598,15 +598,14 @@ def migrate_l3_routers_from_agent(qclient, agent, targets, agent_picker,
                                   delete_namespace):
     LOG.info("Querying agent_id=%s for routers to migrate away", agent['id'])
-    router_id_list = list_routers_on_l3_agent(qclient, agent['id'])
-    router_id_list = router_filter.filter_routers(router_id_list)
```
It seems this line has gone fishing. The aim of the router filter is to be able to only migrate a specified set of routers - if we remove that, we might be in trouble as user would expect only some routers are migrated, and all of the routers will be. The other question - why it was not caught by tests - might be valid from your side. That's something that needs to be fixed.

AbelNavarro replied:
very nice catch. I won't change the router filter to deal with routers instead of router ids since it writes the ids to a file. Instead I'll use some list comprehensions to get the same behavior.

matelakat reviewed (changes requested):
@abel-navarro I think the router filter functionality would break with this change.

matelakat reviewed (approved):
Thanks @abel-navarro !
